Feature/chad 12650/md exploring

matthewdehaven opened the pull request.

matthewdehaven committed:
Add a setpoint_bounds.lua script and load it

This file is intended to contain code for dynamically retrieving setpoint constraints from zwave thermostats. The initial contents are copied from a zigbee bulb color bounds script with a similar purpose.

matthewdehaven committed:
Add beginning of setpoint_bounds object and return in

matthewdehaven committed:
Create a factory method for setpoint capabilities handler

matthewdehaven committed:
Add a placeholder zwave thermostat capabilites report handler

matthewdehaven committed:
Add arguments to setpoint capabilites report handler

matthewdehaven committed:
Request setpoint capabilites from capatible zwave thermostats

matthewdehaven committed:
Fix typo: double "capabilities"

matthewdehaven committed:
Use V3 Thermostat Setpoint Command

matthewdehaven committed:
Update test case to expect Capabilities Get command send

matthewdehaven committed:
Add send for Cooling Setpoint Capabilities Get

matthewdehaven committed:
Add start of setpoint capability report test

matthewdehaven committed:
Add a fake return value for heating setpoint min/max

matthewdehaven committed:
Emit placeholder event for setpoint report

matthewdehaven committed:
Convert F to C

matthewdehaven committed:
Use variables instead of hardcoded values

matthewdehaven committed:
Implement setpoint report conversion

matthewdehaven committed:
Remove a FIXME

matthewdehaven committed:
Adjust fixme comment

matthewdehaven committed:
Delete unused file

matthewdehaven committed:
Remove reference to unused file

greens commented on drivers/SmartThings/zwave-thermostat/src/init.lua:
```diff
@@ -87,6 +102,32 @@ local function set_setpoint_factory(setpoint_type)
+  local min_temp_c = celsius_setpoint_temperature(args.precision1, args.min_value, args.scale1)
+  local max_temp_c = celsius_setpoint_temperature(args.precision2, args.max_value, args.scale2)
+
+  device:emit_event_for_endpoint(cmd.src_channel, capabilities.thermostatHeatingSetpoint.heatingSetpointRange(
```
so which capability you send this event for (`thermostatHeatingSetpoint` or `thermostatCoolingSetpoint`) will depend on `cmd.args.setpoint_type`. Specifically, we'll only be looking for `ThermostatSetpoint.setpoint_type.HEATING_1` and `ThermostatSetpoint.setpoint_type.HEATING_2`

matthewdehaven replied:
Do you mean `ThermostatSetpoint.setpoint_type.COOLING_1` instead of `ThermostatSetpoint.setpoint_type.HEATING_2`? I don't see a reference to HEATING_2 anywhere.

greens replied:
I did in fact mean that.

greens commented on drivers/SmartThings/zwave-thermostat/src/init.lua:
```diff
@@ -87,6 +102,32 @@ local function set_setpoint_factory(setpoint_type)
+    end
+  end
+  local args = cmd.args
+  --- FIXME - MAD - Verify scale1 is associated with min_value and scale2 is associated with max_value
```
this is correct

greens commented on drivers/SmartThings/zwave-thermostat/src/init.lua:
```diff
@@ -87,6 +102,32 @@ local function set_setpoint_factory(setpoint_type)
 end
 
+local function setpoint_capabilites_report(driver, device, cmd)
+  local function concrete_setpoint_temperature(precision, value)
+    local result = value / (10 ^ precision)
+    return result
+  end
```
we have a little bit of magic that we do in our generated z-wave libraries so that fields encoded with precision and size are just automatically translated to numbers, so you should just be able to use `cmd.args.min_value` and `cmd.args.max_value` without having to do the manual precision calculation.

greens commented on drivers/SmartThings/zwave-thermostat/src/test/test_zwave_thermostat.lua:
```diff
@@ -379,6 +396,46 @@ test.register_message_test(
+          ThermostatSetpointV3:CapabilitiesReport(
+            {
+              setpoint_type = ThermostatSetpoint.setpoint_type.HEATING_1,
+              size1 = 4,
+              scale1 = ThermostatSetpoint.scale.CELSIUS,
+              precision1 = 2,
+              min_value = 722,
```
part of that magic includes the serialization, so you _should_ be able to get by with just omitting `sizeX` and `precisionX` and just putting `72.2` for the `min_value` and it should "just work"

matthewdehaven committed:
Utitlize implicit serialization magic

Do not explicitly use scale and precision. Rely on existing conversion utilities behind the scenes.

matthewdehaven committed:
Remove a fixme comment

matthewdehaven committed:
Improve specificity of test name

matthewdehaven committed:
Add heating Fahrenheit test

and fix an issue identified by this new test

matthewdehaven committed:
Copy heating tests to create cooling test

matthewdehaven committed:
heating -> cooling in copied test

matthewdehaven committed:
Support cooling setpoint range report

greens commented on drivers/SmartThings/zwave-thermostat/src/init.lua:
```diff
@@ -87,6 +102,31 @@ local function set_setpoint_factory(setpoint_type)
+  local capability_constructor = nil
+  if args.setpoint_type == ThermostatSetpoint.setpoint_type.HEATING_1 then
+    capability_constructor = capabilities.thermostatHeatingSetpoint.heatingSetpointRange
+  else
```
you should specifically check for `COOLING_1` since it is possible (if unlikely) that some other controller on the network inquires about the limits for one of the other setpoint types we don't support.

greens commented on drivers/SmartThings/zwave-thermostat/src/init.lua:
```diff
@@ -87,6 +102,31 @@ local function set_setpoint_factory(setpoint_type)
 
+local function setpoint_capabilites_report(driver, device, cmd)
+  local args = cmd.args
+  local min_temp_c = args.min_value
+  local max_temp_c = args.max_value
```
rename these, since they're not guaranteed to be in C

matthewdehaven committed:
Rename min/max temp variables

They are not guaranteed to be in celsius

matthewdehaven committed:
Explicitly check for cooling setpoint

To account for the case a different controller on the network inquires about an unsupported setpoint

matthewdehaven commented:
@greens can I open a pull request on the main repo? I created this fork here because I was just exploring the code, but it seems to me it may be good to move over to the main repo now.

greens commented:
@matthewdehaven Yep. Go ahead! I think you could actually just change the target of this PR to be the main branch of the origin repo, or you can open a new PR.

matthewdehaven commented:
@greens 

I couldn't change the target repository of this merge request, so I just opened a new one here: https://github.com/SmartThingsCommunity/SmartThingsEdgeDrivers/pull/1344

I can't request reviewers, though. I'm guessing I'm not a member of the correct group.

I need some help with review request part of the PR process.
